app/ledger.js

```diff
@@ -0,0 +1,915 @@
+    if (synopsis.publishers[publisher].scores[scorekeeper] <= 0) return
+
+    results.push(underscore.extend({ publisher: publisher }, underscore.omit(synopsis.publishers[publisher], 'window')))
+  }, synopsis)
+  results = underscore.sortBy(results, (entry) => { return -entry.scores[scorekeeper] })
+  n = results.length
+
+  total = 0
+  for (i = 0; i < n; i++) { total += results[i].scores[scorekeeper] }
+  if (total === 0) return data
+
+  pct = []
+  for (i = 0; i < n; i++) {
+    publisher = synopsis.publishers[results[i].publisher]
+    duration = results[i].duration
+
+    data[i] = { rank: i + 1,
+                site: results[i].publisher, views: results[i].visits, duration: duration,
+                daysSpent: 0, hoursSpent: 0, minutesSpent: 0, secondsSpent: 0,
+                faviconURL: publisher.faviconURL
+              }
+    if (results[i].protocol) data[i].publisherURL = results[i].protocol + '//' + results[i].publisher
+    // TBD: remove after post-beta... [MTR]
+    if (!data[i].publisherURL) data[i].publisherURL = 'http://' + results[i].publisher
+
+    pct[i] = Math.round((results[i].scores[scorekeeper] * 100) / total)
+
+    if (duration >= msecs.day) {
+      data[i].daysSpent = Math.max(Math.round(duration / msecs.day), 1)
+    } else if (duration >= msecs.hour) {
+      data[i].hoursSpent = Math.max(Math.floor(duration / msecs.hour), 1)
+      data[i].minutesSpent = Math.round((duration % msecs.hour) / msecs.minute)
+    } else if (duration >= msecs.minute) {
+      data[i].minutesSpent = Math.max(Math.round(duration / msecs.minute), 1)
+      data[i].secondsSpent = Math.round((duration % msecs.minute) / msecs.second)
+    } else {
+      data[i].secondsSpent = Math.max(Math.round(duration / msecs.second), 1)
+    }
+  }
+
+  // courtesy of https://stackoverflow.com/questions/13483430/how-to-make-rounded-percentages-add-up-to-100#13485888
+  var foo = (l, target) => {
+    var off = target - underscore.reduce(l, (acc, x) => { return acc + Math.round(x) }, 0)
+
+    return underscore.chain(l)
+                     .sortBy((x) => { return Math.round(x) - x })
+                     .map((x, i) => { return Math.round(x) + (off > i) - (i >= (l.length + off)) })
+                     .value()
+  }
+
+  pct = foo(pct, 100)
+  total = 0
+  for (i = 0; i < n; i++) {
+/*
+    if (pct[i] <= 0) {
+      data = data.slice(0, i)
+      break
+    }
+ */
+    if (pct[i] < 0) pct[i] = 0
+
+    data[i].percentage = pct[i]
+    total += pct[i]
+  }
+
+  for (i = data.length - 1; (total > 100) && (i >= 0); i--) {
+    if (data[i].percentage < 2) continue
+
+    data[i].percentage--
+    total--
+  }
+
+  return data
+}
+
+/*
+ * publisher utilities
+ */
+
+var currentLocation = 'NOOP'
+var currentTimestamp = underscore.now()
+
+var visit = (location, timestamp) => {
+  var setLocation = () => {
+    var duration, publisher
+
+    if (location !== currentLocation) console.log('new location: ' + location)
```

Comment:
Sure!  Feel free to remove it... Just there for debug!

/mtr




On Mon, Aug 15, 2016 at 4:51 PM -0700, "yan zhu (@bcrypt)" <notifications@github.com> wrote:












In app/ledger.js:
> +
> +  return data
> +}
> +
> +/*
> + * publisher utilities
> + */
> +
> +var currentLocation = 'NOOP'
> +var currentTimestamp = underscore.now()
> +
> +var visit = (location, timestamp) => {
> +  var setLocation = () => {
> +    var duration, publisher
> +
> +    if (location !== currentLocation) console.log('new location: ' + location)



@mrose17 can we remove this console.log? URLs from private browsing should not be logged



—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub, or mute the thread.